File: Algorithms-C/Divide and Conquer/MergeSort.c
#include<stdio.h>
int n;

void printArray(int arr[]){
	printf("Displaying the array\n");
	for(int i=0;i<n;i++){
		printf("%d ",arr[i]);
	}
	printf("\n");
}

void merger(int arr[], int start, int mid, int end){
	int len = end - start;
	int mix[len];
	int i = start;
	int j = mid;
	int k = 0;
	while(i < mid && j < end){
		if(arr[i] < arr[j]){
			mix[k] = arr[i];
			i++;
		}
		else{
			mix[k] = arr[j];
			j++;
		}
		k++;
	}
	while(i < mid){
		mix[k] = arr[i];
		k++;i++;
	}	
	while(j < end){
		mix[k] = arr[j];
		k++;j++;
	}
	for(int l = 0;l<len;l++){
		arr[start + l] = mix[l];
	}
}

void mergeSort(int arr[], int start, int end){
	if(end - start == 1){
		return;
	}
	int mid = start + (end - start)/2;
	mergeSort(arr,start,mid);
	mergeSort(arr,mid,end);
	
	merger(arr,start,mid,end);
}

void main(){
	printf("Enter the number of elements\n");
	scanf("%d",&n);
	int arr[n];
	printf("Enter the array elements\n");
	for(int i=0;i<n;i++){
		scanf("%d",&arr[i]);
	}
	mergeSort(arr,0,n);
	printArray(arr);
}

File: Python/Practice/MergeSort.py
def merge(list,s,mid,e):
    len = e - s
    mix = []
    i = s
    j = mid

    while i < mid and j < e:
        if list[i] < list[j]:
            mix.append(list[i])
            i += 1
        else:
            mix.append(list[j])
            j += 1

    while i < mid:
        mix.append(list[i])
        i += 1

    while j < e:
        mix.append(list[j])
        j += 1
    
    for l in range(len):
        list[s + l] = mix[l]

def mergesort(list,s,e):
    if e == s + 1:
        return

    mid = s + (e - s) // 2
    
    mergesort(list,s,mid)
    mergesort(list,mid,e)
    
    merge(list,s,mid,e)

    return


# list = [2,1,4,5,3,6,7]
list = [23,12,56,7,34,67]
mergesort(list,0,len(list))
print(list)